numba/np/arrayobj.py

```diff
@@ -5168,80 +5168,72 @@ def _normalize_axis(context, builder, func_name, ndim, axis):
     return axis
 
 
-def _insert_axis_in_shape(context, builder, orig_shape, ndim, axis):
-    """
-    Compute shape with the new axis inserted
-    e.g. given original shape (2, 3, 4) and axis=2,
-    the returned new shape is (2, 3, 1, 4).
-    """
-    assert len(orig_shape) == ndim - 1
-
-    ll_shty = ir.ArrayType(cgutils.intp_t, ndim)
-    shapes = cgutils.alloca_once(builder, ll_shty)
-
-    one = cgutils.intp_t(1)
-
-    # 1. copy original sizes at appropriate places
-    for dim in range(ndim - 1):
-        ll_dim = cgutils.intp_t(dim)
-        after_axis = builder.icmp_signed('>=', ll_dim, axis)
-        sh = orig_shape[dim]
-        idx = builder.select(after_axis,
-                             builder.add(ll_dim, one),
-                             ll_dim)
-        builder.store(sh, cgutils.gep_inbounds(builder, shapes, 0, idx))
-
-    # 2. insert new size (1) at axis dimension
-    builder.store(one, cgutils.gep_inbounds(builder, shapes, 0, axis))
-
-    return cgutils.unpack_tuple(builder, builder.load(shapes))
-
-
-def _insert_axis_in_strides(context, builder, orig_strides, ndim, axis):
-    """
-    Same as _insert_axis_in_shape(), but with a strides array.
-    """
-    assert len(orig_strides) == ndim - 1
-
-    ll_shty = ir.ArrayType(cgutils.intp_t, ndim)
-    strides = cgutils.alloca_once(builder, ll_shty)
-
-    one = cgutils.intp_t(1)
-    zero = cgutils.intp_t(0)
-
-    # 1. copy original strides at appropriate places
-    for dim in range(ndim - 1):
-        ll_dim = cgutils.intp_t(dim)
-        after_axis = builder.icmp_signed('>=', ll_dim, axis)
-        idx = builder.select(after_axis,
-                             builder.add(ll_dim, one),
-                             ll_dim)
-        builder.store(orig_strides[dim],
-                      cgutils.gep_inbounds(builder, strides, 0, idx))
-
-    # 2. insert new stride at axis dimension
-    # (the value is indifferent for a 1-sized dimension, we use 0)
-    builder.store(zero, cgutils.gep_inbounds(builder, strides, 0, axis))
-
-    return cgutils.unpack_tuple(builder, builder.load(strides))
-
-
-def expand_dims(context, builder, sig, args, axis):
+def expand_dims(context, builder, sig, args):
     """
     np.expand_dims() with the given axis.
     """
     retty = sig.return_type
     ndim = retty.ndim
     arrty = sig.args[0]
+    _empty_tuple = tuple([-1] * ndim)
+
+    if isinstance(sig.args[1], types.Integer):
+        to_tuple = register_jitable(lambda axis: (axis,))
+    else:
```

Comment:
The documentation on https://numpy.org/doc/stable/reference/generated/numpy.expand_dims.html seems to suggest `axis` can either be an int or a tuple of ints. In this case however it could also be an array or a string?

Reply:
The validity of arguments is being checked at type level:

https://github.com/numba/numba/pull/7604/files#diff-d0aefd08783016c4e2a19e29d4725c7870806ea2aa1ab4caa6e8f0adc08c6e64R5287-R5293